src/darwin/Framework/CHIP/MTRAttributeValueWaiter.mm

```diff
@@ -120,15 +137,45 @@ - (void)_notifyWithError:(NSError * _Nullable)error
         MTR_LOG("%@ %p wait for attribute values unknown error: %@", self, self, error);
     }
 
-    auto completion = self.completion;
-    auto queue = self.queue;
-    self.completion = nil;
-    self.queue = nil;
+    auto * device = self.device;
     dispatch_async(queue, ^{
+        // Call into the device from our async dispatch, to avoid problems when
+        // _notifyWithError is happening while the device lock is held.
+        //
+        // But this seems to trigger TSAN errors about null-derefs.  Probably
+        // when this code is called from dealloc, in which case self cannot
+        // really be captured, since it's dying.
+        [device _forgetAttributeWaiter:self];
         completion(error);
     });
 }
 
+- (void)_startTimer:(NSTimeInterval)timeout withQueue:(dispatch_queue_t)queue
+{
+    dispatch_source_t timerSource = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, queue);
```

Comment:
Using `self.device.queue` may be more obvious than passing in the queue? Could also consider doing this directly in `init`?

Reply:
Doing it in init would in some cases start and immediately stop the timer because our values are in fact satisfied.  So I'd rather not start in init.

`self.device.queue` I can do.